File: lib/controllers/join.js
const _    = require('lodash/fp');
const join = require('../pairrit').join;
const leave = require('../pairrit').leave;

const firstValueFrom = (array) => {
  return _.first(_.compact(array));
};

const derivePairNameFrom = ({text, user_name}) => {
  const name =  _.split(' ', text)[1];

  return firstValueFrom([name, user_name]);
};

module.exports = (req, res) => {
  const userName  = req.body.user_name;
  const pairName  = derivePairNameFrom(req.body);
  const channelId = req.body.channel_id;

  leave(userName, channelId)
    .then(() => {
      return join(userName, pairName, channelId);
    })
    .then(() => {
      res.send(`Welcome to the \`${pairName}\` pair!`);
    });
  // if we calculate the key sooner we can do
  // const result = join(userName, pairKey);

};

File: app/models/command/join.rb
class Command::Join < Command::Base
  attr_reader :channel, :params, :user

  def initialize(params)
    @params = params
    @channel = params['channel']
    @user = params['user']
  end

  def process
    user.leave(channel)

    pair = Pair.find_or_create_by(channel: channel, name: pair_name)
    pair.add(user)

    { text: "Welcome to the `#{pair.name}` pair!" }
  end

  private

  def pair_name
    name = params['user_name']

    if params['args'].any?
      name = params['args'].first
    end

    name
  end
end
